<h1><center>Tarea 1: Roja o Azul</center></h1>

<center><strong>MDS7202: Laboratorio de Programación Científica para Ciencia de Datos</strong></center>

### Cuerpo Docente:

- Profesor: Pablo Badilla
- Auxiliar: Ignacio Meza D.
- Ayudante: Diego Irarrázaval

### Equipo: SUPER IMPORTANTE - notebooks sin nombre no serán revisados

- Nombre de alumno 1: Joaquín Roa
- Nombre de alumno 2: Matías Sepúlveda


### **Link de repositorio de GitHub:** `https://github.com/jroallanos/LDPCPCDD19`

# Temas a tratar

- Programación orientada a objetos.
- Manejo de datos en arreglos/tensores con `Numpy`.
- Análisis exploratorio de datos con `Pandas`
- Preprocesamiento con `Scikit-learn`.
- Composición de preprocesadores con Pipelines.

## Reglas:

- Fecha de entrega: Descrita en U-cursos
- **Grupos de 2 personas**
- Cualquier duda fuera del horario de clases al foro. Mensajes al equipo docente serán respondidos por este medio.
- Prohibidas las copias. 
- Pueden usar cualquier material del curso que estimen conveniente.



# Descripción del tarea (LEER ESTO).

En el presente notebook se presentan dos problemas, uno de **Segmentación de clientes en tiendas de retail** y otro enfocado en el **Análisis de temperaturas**. Producto de la larga extensión que podría significar la realización de los dos problemas, para efectos de esta tarea deben escoger **SOLAMENTE UNA** sección para desarrollar. Si desea puede borrar la pregunta que no realizará.

<p align="center">
  <img src="https://media1.giphy.com/media/ikwWGan6JZMo7bbf5L/giphy.gif?cid=790b761140b82e1c3d6754790ef49d96e8406bdac2f36d6f&rid=giphy.gif&ct=g">
</p>




#Importamos librerias utiles 😸

In [1]:
# Libreria Core del lab.
import numpy as np
import pandas as pd
import datetime
from IPython.display import HTML

# Libreria para plotear (En colab esta desactualizado plotly)
!pip install --upgrade plotly
import plotly.express as px

# Librerias utiles
from sklearn.preprocessing import StandardScaler
from sklearn.cluster import KMeans
from sklearn.base import BaseEstimator, TransformerMixin
from sklearn.tree import DecisionTreeRegressor
from sklearn.model_selection import train_test_split 
from sklearn.metrics import r2_score 
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import MinMaxScaler
from sklearn.preprocessing import FunctionTransformer
from sklearn.preprocessing import PowerTransformer
from sklearn.preprocessing import OneHotEncoder
import calendar

zsh:1: command not found: pip


# 2 Clima Chileno ☀️

In [2]:
HTML('<p align="center"><iframe width="560" height="315" src="https://www.youtube.com/embed/fGr-njYF_U0?rel=0&amp;controls=0&amp;showinfo=0" frameborder="0" allowfullscreen></iframe></p>')

## 2.1 Cargar Dataset

<p align="center">
  <img src="https://cdn.worldvectorlogo.com/logos/kiwi-3.svg" width='150px'>
</p>

`Kiguix`, Una prestigiosa empresa de software esta interesada en ingresar al mundo de la predicción del tiempo, por lo que se ha puesto en contacto con ustedes (unos expertos en Data Science) para analizar los datos.

Para comenzar el administrador de ```Kiguix``` les entrega el dataset llamado ```cr2_tasDaily_2018.csv``` el que posee un gran registro de temperaturas obtenidas en mas de **300 estaciones metereologicas*** de Chile. Con esto les solicita si pueden explorar este dataset y ver si es posible observar caracteristicas relevantes de los datos.

Como **primera Task** se le solicita cargar el archivo y mostrar una pequeña muestra del dataset a estudiar.

In [3]:
# Primero vamos a cargar los datos del tiempo y luego mostraremos unas muestras de ejemplo
#########################
temp_chile = pd.read_csv("/Users/jroallanos/Desktop/Laboratorio de Programación Científica para Ciencia de Datos/tarea 1/cr2_tasDaily_2018.csv")
temp_chile
#########################

/opt/anaconda3/lib/python3.8/site-packages/IPython/core/interactiveshell.py:3071: DtypeWarning: Columns (1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63,64,65,66,67,68,69,70,71,72,73,74,75,76,77,78,79,80,81,82,83,84,85,86,87,88,89,90,91,92,93,94,95,96,97,98,99,100,101,102,103,104,105,106,107,108,109,110,111,112,113,114,115,116,117,118,119,120,121,122,123,124,125,126,127,128,129,130,131,132,133,134,135,136,137,138,139,140,141,142,143,144,145,146,147,148,149,150,151,152,153,154,155,156,157,158,159,160,161,162,163,164,165,166,167,168,169,170,171,172,173,174,175,176,177,178,179,180,181,182,183,184,185,186,187,188,189,190,191,192,193,194,195,196,197,198,199,200,201,202,203,204,205,206,207,208,209,210,211,212,213,214,215,216,217,218,219,220,221,222,223,224,225,226,227,228,229,230,231,232,233,234,235,236,237,238,239,240,241,242,243,244,245,246,247,248,249,250,251

codigo_estacion 01000005                     01201005  01001005  \
0         institucion      DGA                          DGA       DGA   
1              fuente  dga_web                      dga_web   dga_web   
2              nombre  Visviri  Rio Caracarani En Humapalca   Caquena   
3              altura     4080                         3908      4400   
4             latitud  -17.595                     -17.8428  -18.0542   
...               ...      ...                          ...       ...   
43176      2018-03-05    6.822                      -9999.0     -9999   
43177      2018-03-06    5.896                      -9999.0     -9999   
43178      2018-03-07    6.562                      -9999.0     -9999   
43179      2018-03-08    6.887                      -9999.0     -9999   
43180      2018-03-09    6.962                      -9999.0     -9999   

       01202009     01202012  01202010              01020017        01010010  \
0           DGA          DGA       DGA                   DGA             DGA   
1       dga_web      dga_web   dga_web               dga_web         dga_web   
2       Pacollo  Putre (Dcp)     Putre  Parinacota Ex Endesa  Chungara Ajata   
3          4185         3560      3545                  4420            4585   
4      -18.1769      -18.195  -18.1992              -18.2042        -18.2353   
...         ...          ...       ...                   ...             ...   
43176     -9999        2.865     -9999                 3.918           3.848   
43177     -9999        2.647     -9999                 3.774           3.738   
43178     -9999        3.136     -9999                 4.584            4.33   
43179     -9999        3.193     -9999                 4.258           4.543   
43180     -9999        3.482     -9999                 4.043            4.71   

         01300004  ...             330021                        330019  \
0             DGA  ...                DMC                           DMC   
1         dga_web  ...            dmc_web                       dmc_web   
2      Murmuntane  ...  Pudahuel Santiago  Eulogio Sanchez Tobalaba Ad.   
3            3550  ...                482                           650   
4        -18.3519  ...           -33.3919                      -33.4544   
...           ...  ...                ...                           ...   
43176       -9999  ...               22.3                          21.5   
43177       -9999  ...               19.9                       -9999.0   
43178       -9999  ...               17.8                          18.8   
43179       -9999  ...               18.3                          18.4   
43180       -9999  ...               18.4                       -9999.0   

                                      330031             330030  \
0                                        DMC                DMC   
1                                    dmc_web            dmc_web   
2      Juan Fernandez Estacion Meteorologica  Santo Domingo Ad.   
3                                         35                 75   
4                                   -33.6367            -33.655   
...                                      ...                ...   
43176                                   18.5            -9999.0   
43177                                   18.4            -9999.0   
43178                                   19.3            -9999.0   
43179                                   19.7               15.0   
43180                                   18.7               14.7   

                              370033       420004         430002  \
0                                DMC          DMC            DMC   
1                            dmc_web      dmc_web        dmc_web   
2      Maria Dolores Los Angeles Ad.  Chaiten Ad.  Futaleufu Ad.   
3                                120           10            347   
4                           -37.4028     -42.9303       -43.1889   
...                              ...          ...     

## 2.2 Exploración y limpieza de datos

Cargado los datos, el administrador les consulta si es posible realizar un analisis de los datos en el estado que se encuentran. ¿Es posible observar información relevante al realizar un analisis exploratorio de los datos?. Recuerde justificar su respuesta en la casilla **Justifique**, utilice como argumento lo visto en la exploración de datos realizada.

### 2.2.1 ¿Es posible realizar en este estado un análisis exploratorio de los datos? [0,5]


In [4]:
#########################
def analisis_preliminar(temp_chile):
  display(temp_chile.head())
  print('')
  print("El dataset a estudiar tiene", temp_chile.shape[0],"filas y", temp_chile.shape[1],"columnas.")
  print('')
  print('Revisamos la primera columna: ')
  display(list(temp_chile["codigo_estacion"])[0:20])
  print('')
  print('Investigamos la última fila del dataset: ')
  display(temp_chile.T[43180]) #investigamos la ultima fila del dataset
#########################
analisis_preliminar(temp_chile)

codigo_estacion 01000005                     01201005  01001005  01202009  \
0     institucion      DGA                          DGA       DGA       DGA   
1          fuente  dga_web                      dga_web   dga_web   dga_web   
2          nombre  Visviri  Rio Caracarani En Humapalca   Caquena   Pacollo   
3          altura     4080                         3908      4400      4185   
4         latitud  -17.595                     -17.8428  -18.0542  -18.1769   

      01202012  01202010              01020017        01010010    01300004  \
0          DGA       DGA                   DGA             DGA         DGA   
1      dga_web   dga_web               dga_web         dga_web     dga_web   
2  Putre (Dcp)     Putre  Parinacota Ex Endesa  Chungara Ajata  Murmuntane   
3         3560      3545                  4420            4585        3550   
4      -18.195  -18.1992              -18.2042        -18.2353    -18.3519   

   ...             330021                        330019  \
0  ...                DMC                           DMC   
1  ...            dmc_web                       dmc_web   
2  ...  Pudahuel Santiago  Eulogio Sanchez Tobalaba Ad.   
3  ...                482                           650   
4  ...           -33.3919                      -33.4544   

                                  330031             330030  \
0                                    DMC                DMC   
1                                dmc_web            dmc_web   
2  Juan Fernandez Estacion Meteorologica  Santo Domingo Ad.   
3                                     35                 75   
4                               -33.6367            -33.655   

                          370033       420004         430002          430004  \
0                            DMC          DMC            DMC             DMC   
1                        dmc_web      dmc_web        dmc_web         dmc_web   
2  Maria Dolores Los Angeles Ad.  Chaiten Ad.  Futaleufu Ad.  Alto Palenaad.   
3                            120           10            347             281   
4                       -37.4028     -42.9303       -43.1889        -43.6117   

                                 510005                         530005  
0                                   DMC                            DMC  
1                               dmc_web                        dmc_web  
2  Teniente Gallardo Puerto Natales Ad.  Fuentes Martinez Porvenir Ad.  
3                                    69                             27  
4                              -51.6672                       -53.1875  

[5 rows x 317 columns]


El dataset a estudiar tiene 43181 filas y 317 columnas.

Revisamos la primera columna: 


['institucion',
 'fuente',
 'nombre',
 'altura',
 'latitud',
 'longitud',
 'codigo_cuenca',
 'nombre_cuenca',
 'codigo_sub_cuenca',
 'nombre_sub_cuenca',
 'inicio_observaciones',
 'fin_observaciones',
 'cantidad_observaciones',
 'inicio_automatica',
 '1900-01-01',
 '1900-01-02',
 '1900-01-03',
 '1900-01-04',
 '1900-01-05',
 '1900-01-06']


Investigamos la última fila del dataset: 


codigo_estacion    2018-03-09
01000005                6.962
01201005              -9999.0
01001005                -9999
01202009                -9999
                      ...    
420004                  -9999
430002                -9999.0
430004                -9999.0
510005                    2.6
530005                    5.9
Name: 43180, Length: 317, dtype: object

**Justificación**

```
Puede verse que el nombre de las verdaderas variables (o mejor dicho, de los verdaderos atributos) está realmente contenido en la primera columna del dataset. El resto de las columnas son la observación para cada una de las 316 estaciones metereológicas. 
Esto da una intuición de "transponer" el dataset. Sin embargo, puede verse más adelante que no es directa la transposición, ya que, en primer lugar, son muchas filas (43181) y en la primera columna (se llama "codigo_estacion", aunque no es un buen nombre) hay 14 atributos y el resto de las filas representan una fecha para cada centro de observación. Para las filas correspondientes a una fecha se registra la temperatura que hubo para ese día, desde el año 1900 hasta el 9 de marzo de 2018. 
Puede verse también que existe una gran cantidad de temperaturas mal registradas, que quedan con un valor de -9999. Esto es menor al cero absoluto así que definitivamente es un mal registro.
Por lo anterior mencionado, se requiere una transformación profunda de la estructura del conjunto de datos.
```

### 2.2.2 Limpieza de datos [1 Punto]

Omitiendo su respuesta anterior, el jefe de ```Kiguix``` les comenta que está interesado en generar un Dataset "ordenado" para las estaciones meteorológicas que llevan por **Nombre**:

- 'Pichoy Valdivia Ad.'
- 'Pudahuel Santiago'
- 'Santo Domingo Ad.'
- 'Juan Fernandez Estacion Meteorologica'
- 'Futaleufu Ad.'
- 'Santa Juana'
- 'Lautaro Embalse'

Con esto les solicita que creen un DataFrame con columnas que contengan (1) el año en que fue tomada la muestra, (2) el mes en que se tomó la muestra, (3) la fecha del año y mes en que se tomó la muestra, (4) altura a nivel del mar que se encuentra la estación, (5) latitud y (6) longitud de la estación, (7) el nombre de la estación meteorológica y (8) la temperatura promedio del mes. Su jefe es crítico en señalar que necesita las columnas en este orden y que deja en usted el cargo de tomar las asunciones de que valores omitir en el DataFrame. En caso de ser necesario justifique asunciones.



**Ejemplo de tabla solicitada:**

|      |   Year | Month     | Dates               |   altura |   latitud |   longitud | station                               |      Temp |
|-----:|-------:|:----------|:--------------------|---------:|----------:|-----------:|:--------------------------------------|----------:|
|    0 |   1961 | August    | 1961-08-01 00:00:00 |      347 |  -43.1889 |   -71.8525 | Futaleufu Ad.                         |  6.2      |
|    1 |   1961 | December  | 1961-12-01 00:00:00 |      347 |  -43.1889 |   -71.8525 | Futaleufu Ad.                         | 13        |
|    2 |   1963 | April     | 1963-04-01 00:00:00 |       35 |  -33.6367 |   -78.835  | Juan Fernandez Estacion Meteorologica | 17.9      |
|    3 |   1965 | January   | 1965-01-01 00:00:00 |      560 |  -28.6672 |   -70.6625 | Santa Juana                           | 18.0818   |
|    4 |   1965 | February  | 1965-02-01 00:00:00 |      347 |  -43.1889 |   -71.8525 | Futaleufu Ad.                         | 14.8      |
|    5 |   1965 | February  | 1965-02-01 00:00:00 |      560 |  -28.6672 |   -70.6625 | Santa Juana                           | 21.8214   |
|    6 |   1965 | March     | 1965-03-01 00:00:00 |      560 |  -28.6672 |   -70.6625 | Santa Juana                           | 20.7759   |
|    7 |   1965 | April     | 1965-04-01 00:00:00 |      560 |  -28.6672 |   -70.6625 | Santa Juana                           | 18.244    |
|    8 |   1965 | May       | 1965-05-01 00:00:00 |      560 |  -28.6672 |   -70.6625 | Santa Juana                           | 15.7233   |
|    9 |   1965 | June      | 1965-06-01 00:00:00 |      560 |  -28.6672 |   -70.6625 | Santa Juana                           | 16.1467   |
|   10 |   1965 | July      | 1965-07-01 00:00:00 |      560 |  -28.6672 |   -70.6625 | Santa Juana                           | 14.4516   |

In [5]:
# Primero vamos a selecionar solamente las estaciones que nos interesan
#########################
temp_chile2 = temp_chile.T
temp_chile2 = temp_chile2.reset_index()
temp_chile2.columns = temp_chile2.iloc[0]
temp_chile2 = temp_chile2.drop(0, axis=0)
temp_chile2 = temp_chile2[temp_chile2['nombre'].isin(['Pichoy Valdivia Ad.', 'Pudahuel Santiago', 'Santo Domingo Ad.', 'Juan Fernandez Estacion Meteorologica',
                                                   'Futaleufu Ad.', 'Santa Juana', 'Lautaro Embalse'])]
temp_chile2.head(3)

0   codigo_estacion institucion     fuente               nombre altura  \
66         03430006         DGA    dga_web      Lautaro Embalse   1110   
72         03820004         DGA    dga_web          Santa Juana    560   
294          390006         DMC  dmc_merge  Pichoy Valdivia Ad.     21   

0     latitud  longitud codigo_cuenca nombre_cuenca codigo_sub_cuenca  ...  \
66   -27.9783  -70.0033            34   Rio Copiapo               343  ...   
72   -28.6672  -70.6625            38    Rio Huasco               382  ...   
294  -39.6506  -73.0808           101  Rio Valdivia              1010  ...   

0   2018-02-28 2018-03-01 2018-03-02 2018-03-03 2018-03-04 2018-03-05  \
66      19.794     20.125       22.1     20.971     21.672     19.611   
72      19.266      20.03     20.528     19.928     19.747     18.719   
294       14.0       16.8       15.4       14.3       15.0       14.1   

0   2018-03-06 2018-03-07 2018-03-08 2018-03-09  
66      19.317     21.743     19.479     21.995  
72      17.772     18.743     18.053     20.453  
294       16.6       17.2       15.3       12.9  

[3 rows x 43182 columns]

In [6]:
# Creamos el dataframe de temperaturas donde haremos los cambios necesarios para obtener lo que se pide
temperaturas = temp_chile2.drop(['codigo_estacion', 'altura', 'latitud', 'longitud','institucion', 'fuente', 'codigo_cuenca', 'nombre_cuenca', 
                                   'codigo_sub_cuenca', 'nombre_sub_cuenca', 'inicio_observaciones', 'fin_observaciones', 'cantidad_observaciones', 
                                   'inicio_automatica'], axis=1)   # eliminamos las columnas que no necesitaremos y dejamos la estación + fechas
temperaturas.index = temperaturas['nombre'] # Asignamos por comodidad el nombre de la estación al índice
temperaturas = temperaturas.drop(['nombre'], axis=1) # Eliminamos la columna nombre
temperaturas.head(3)

0                   1900-01-01 1900-01-02 1900-01-03 1900-01-04 1900-01-05  \
nombre                                                                       
Lautaro Embalse          -9999      -9999      -9999      -9999      -9999   
Santa Juana              -9999      -9999      -9999      -9999      -9999   
Pichoy Valdivia Ad.      -9999      -9999      -9999      -9999      -9999   

0                   1900-01-06 1900-01-07 1900-01-08 1900-01-09 1900-01-10  \
nombre                                                                       
Lautaro Embalse          -9999      -9999      -9999      -9999      -9999   
Santa Juana              -9999      -9999      -9999      -9999      -9999   
Pichoy Valdivia Ad.      -9999      -9999      -9999      -9999      -9999   

0                    ... 2018-02-28 2018-03-01 2018-03-02 2018-03-03  \
nombre               ...                                               
Lautaro Embalse      ...     19.794     20.125       22.1     20.971   
Santa Juana          ...     19.266      20.03     20.528     19.928   
Pichoy Valdivia Ad.  ...       14.0       16.8       15.4       14.3   

0                   2018-03-04 2018-03-05 2018-03-06 2018-03-07 2018-03-08  \
nombre                                                                       
Lautaro Embalse         21.672     19.611     19.317     21.743     19.479   
Santa Juana             19.747     18.719     17.772     18.743     18.053   
Pichoy Valdivia Ad.       15.0       14.1       16.6       17.2       15.3   

0                   2018-03-09  
nombre                          
Lautaro Embalse         21.995  
Santa Juana             20.453  
Pichoy Valdivia Ad.       12.9  

[3 rows x 43167 columns]

In [7]:
# Utilizamos melt para pivotear el set de datos
temperaturas = temperaturas.melt(ignore_index = False).reset_index() # ignore_index = False requiere actualizar pd
temperaturas.columns = ['station', 'Dates', 'Temp']
temperaturas.head(3)

station       Dates   Temp
0      Lautaro Embalse  1900-01-01  -9999
1          Santa Juana  1900-01-01  -9999
2  Pichoy Valdivia Ad.  1900-01-01  -9999

In [8]:
# Ya que utilizaremos la temperatura promedio por mes, para evitar sesgo vamos a ignorar los días con temperaturas que no tienen lógica.
temperaturas = temperaturas[ temperaturas['Temp'] != '-9999'] # Primera forma en que están escritas las temperaturas erroneas o código para mediciones erróneas '-9999'
temperaturas = temperaturas[ temperaturas['Temp'] != -9999].reset_index(drop=True) # Segunda forma en que están escritas, -9999
temperaturas['Year'] = pd.DatetimeIndex(temperaturas['Dates']).year # Creamos la columna con el año al que corresponde la medición.
temperaturas['Month'] = pd.DatetimeIndex(temperaturas['Dates']).month # Creamos la columna con el mes al que corresponde la medición.
temperaturas = temperaturas.drop(['Dates'], axis=1) # Eliminamos la columna fecha que ya no vamos a necesitar, pero más adelante crearemos la solicitada.
temperaturas['Temp'] = pd.to_numeric(temperaturas['Temp']) # La serie era de tipo objeto por lo que es necesario actualizar  a numérica.
temperaturas_final = temperaturas.groupby(['station','Year', 'Month']).mean().reset_index() # Se realiza un groupby para obtener la temperatura promedio por las variables dadas
temperaturas_mean = temperaturas.groupby(['Year', 'Month']).mean().reset_index()


In [9]:
temperaturas_final = temperaturas_final.drop(['Temp'], axis=1)
temperaturas = pd.merge(left=temperaturas_final, right = temperaturas_mean, on = ['Year', 'Month'], how = "inner")

In [10]:
#########################
info_station = temp_chile2.loc[:,['nombre', 'altura', 'latitud', 'longitud']] 
info_station.columns = ['station', 'altura', 'latitud', 'longitud']
info_station.head(3)

station altura   latitud  longitud
66       Lautaro Embalse   1110  -27.9783  -70.0033
72           Santa Juana    560  -28.6672  -70.6625
294  Pichoy Valdivia Ad.     21  -39.6506  -73.0808

In [25]:
#########################
temp_chile_ord = pd.merge(left=temperaturas, right=info_station, on = ['station'], how="inner") # Unimos las temperaturas con los datos de la estación.
#########################
temp_chile_ord['Dates'] = 0
for index in range(len(temp_chile_ord)): # Creamos un loop para poder agregar la nueva columna Dates en base al año y mes.
  una_fecha = str(temp_chile_ord['Year'][index]) + '/' + str(temp_chile_ord['Month'][index]) + '/01 00:00:00'
  temp_chile_ord['Dates'][index] = datetime.datetime.strptime(una_fecha, '%Y/%m/%d %H:%M:%S')
#########################
cols = ['Year', 'Month', 'Dates', 'altura', 'latitud', 'longitud', 'station', 'Temp'] # Orden en el que queremos el dataframe
temp_chile_ord = temp_chile_ord[cols]
temp_chile_ord=temp_chile_ord.sort_values(by=['Dates']) # Se ordena el dataframe en función de las fechas
temp_chile_ord=temp_chile_ord.reset_index(drop=True)
months=temp_chile_ord.copy()
temp_chile_ord['Month'] = temp_chile_ord['Month'].apply(lambda x: calendar.month_name[x]) # Cambiamos el mes de manera numerica al nombre correspondiente.

<ipython-input-25-9ce0d4264b06>:7: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



In [12]:
display(temp_chile_ord.head(11))

Year     Month                Dates altura   latitud  longitud  \
0   1961    August  1961-08-01 00:00:00    347  -43.1889  -71.8525   
1   1961  December  1961-12-01 00:00:00    347  -43.1889  -71.8525   
2   1963     April  1963-04-01 00:00:00     35  -33.6367   -78.835   
3   1965   January  1965-01-01 00:00:00    560  -28.6672  -70.6625   
4   1965  February  1965-02-01 00:00:00    560  -28.6672  -70.6625   
5   1965  February  1965-02-01 00:00:00    347  -43.1889  -71.8525   
6   1965     March  1965-03-01 00:00:00    560  -28.6672  -70.6625   
7   1965     April  1965-04-01 00:00:00    560  -28.6672  -70.6625   
8   1965       May  1965-05-01 00:00:00    560  -28.6672  -70.6625   
9   1965      June  1965-06-01 00:00:00    560  -28.6672  -70.6625   
10  1965      July  1965-07-01 00:00:00    560  -28.6672  -70.6625   

                                  station       Temp  
0                           Futaleufu Ad.   6.200000  
1                           Futaleufu Ad.  13.000000  
2   Juan Fernandez Estacion Meteorologica  17.900000  
3                             Santa Juana  18.081818  
4                             Santa Juana  21.579310  
5                           Futaleufu Ad.  21.579310  
6                             Santa Juana  20.775862  
7                             Santa Juana  18.244000  
8                             Santa Juana  15.723333  
9                             Santa Juana  16.146667  
10                            Santa Juana  14.451613

**Justifique:**

```
Algo que es importante de acotar, fuera del código explicado anteriormente, es la decisión de eliminar los datos -9999 que suponemos es la forma que se tiene de guardar un tipo de datos erróneo, en especial dado que posteriormente veremos que la temperatura cae dentro de márgenes coherentes al realizar dicha limpieza de datos.
```

### 2.2.3 Nueva exploración de datos [1 Punto]

Visto que el nuevo DataFrame entrega una estructura mas ordenada de los datos se decide hacer una nueva exploración de los datos, para esto se les pide que:

- Aplicar las 3 técnicas de EDA que usted estime conveniente.
- Realicen un Box-plot temporal de los Datos utilizando plotly.
- Graficar tendencia de la temperatura para cada uno de los meses del año.

Comente que puede observar de los datos. ¿Es posible observar alguna tendencia o patron?. Justifique.

In [13]:
print("El nuevo dataset tiene un total de", temp_chile_ord.shape[0],"filas de observaciones y",temp_chile_ord.shape[1],"columnas")

El nuevo dataset tiene un total de 3387 filas de observaciones y 8 columnas


In [14]:
print("Tabla descriptiva de los datos numéricos")
print('')
display(temp_chile_ord.describe().T)
print('')
print("Puede verse que la mitad de los datos fueron recopilados durante o antes de 1988.")
print("El mínimo año disponible para el análisis es 1961, y el máximo 2018.")
print("Transcurrieron",1988-1961,"años entre la primera observación y la mediana,")
print("y",2018-1988,"años entre la última observación y la mediana.")

Tabla descriptiva de los datos numéricos



count         mean        std     min          25%          50%  \
Year  3387.0  1989.532625  14.870311  1961.0  1977.000000  1988.000000   
Temp  3387.0    13.808550   3.256114     6.2    10.984112    13.773377   

              75%          max  
Year  2001.000000  2018.000000  
Temp    16.653333    22.006557


Puede verse que la mitad de los datos fueron recopilados durante o antes de 1988.
El mínimo año disponible para el análisis es 1961, y el máximo 2018.
Transcurrieron 27 años entre la primera observación y la mediana,
y 30 años entre la última observación y la mediana.


In [15]:
print("Cantidad de registros, en orden descendiente, por estación")
print('')
display(temp_chile_ord.groupby('station', as_index=False).agg({"Temp": "count"}).sort_values(by=["Temp"],ascending=False))
print('')
print("La estación con mayor número de recopilaciones de datos fue la estación de Juan Fernández")
print("y la estación con menor cantidad de registros es Lautaro Embalse.")

Cantidad de registros, en orden descendiente, por estación



station  Temp
1  Juan Fernandez Estacion Meteorologica   625
4                      Pudahuel Santiago   613
3                    Pichoy Valdivia Ad.   529
6                      Santo Domingo Ad.   528
0                          Futaleufu Ad.   468
5                            Santa Juana   338
2                        Lautaro Embalse   286


La estación con mayor número de recopilaciones de datos fue la estación de Juan Fernández
y la estación con menor cantidad de registros es Lautaro Embalse.


In [27]:
fig = px.box(temp_chile_ord[(temp_chile_ord["Year"]>1965)&(temp_chile_ord["Year"]<2018)], x="Year", y="Temp",title="Temperaturas entre 1966 y 2017")
fig.show()

In [17]:
fig2 = px.box(temp_chile_ord[(temp_chile_ord["Year"]>1964)&(temp_chile_ord["Year"]<2018)], x="Month", y="Temp",title="Temperaturas entre 1965 y 2017, por mes")
fig2.show()

In [18]:
import plotly.graph_objects as go
from plotly.subplots import make_subplots
meses = ['January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 
         'September', 'October', 'November', 'December']

fig = make_subplots(
    rows = 3, cols = 4, subplot_titles = meses
)
row_id = 1
col_id = 0
for month in meses:
  if col_id < 4:
    col_id = col_id + 1
    data = temp_chile_ord[temp_chile_ord['Month'] == month]
    fig.add_trace(
        go.Scatter(x = data['Year'], y = data['Temp']),
                   row = row_id , col = col_id 
    )
  else:
    row_id = row_id + 1
    col_id = 1
    data = temp_chile_ord[temp_chile_ord['Month'] == month]
    fig.add_trace(
        go.Scatter(x = data['Year'], y = data['Temp']),
                   row = row_id , col = col_id 
    )


fig.update_layout(height=800,showlegend=False, title_text="Monthly temperature throught history:")
fig.show()

### 2.2.3.1 Representación interesante [Bonus - 0,5 Puntos]


Implemente la siguiente visualización y proponga otra que considere útil.
Visite la referencia de [plotly](https://plotly.com/python) y [px.line](https://plotly.com/python/line-charts/) para mayor información.

In [28]:
fig = px.line(temp_chile_ord[temp_chile_ord['Year']>=1967], x = "Month", y = "Temp", animation_frame = "Year", range_y = [-1, 26],title="Evolución temporal de la temperatura, por mes y año")
fig.show()

In [29]:
#########################

fig = px.line(temp_chile_ord[temp_chile_ord['Year']>=1967], x = "Month", y = "Temp",color="station", animation_frame = "Year", range_y = [-1, 26],title="Evolución temporal de la temperatura, por mes y año, diferenciado por estación")
fig.show()

#########################

## 2.3 Pipeline y Predicción

### 2.3.1 Encoding [2 Puntos]

Contentos con su desempeño ```Kiguix``` les pide realizar su tarea más ambiciosa, la que consiste en crear un pipeline para predecir temperaturas futuras. Para esto les señala que desea realizar el siguiente encoding para las columnas del DataFrame creado en la sección ```2.2.2```:

- Codificar el atributo categórico ```station```.
- Obtener el MinMaxScaler de los atributos ```latitud```, ```longitud``` y ```altura```.
- Realizar una codificación cíclica de ```Year``` y ```Month``` ([referencia de ayuda](https://stats.stackexchange.com/questions/311494/best-practice-for-encoding-datetime-in-machine-learning)).

Para el caso de la codificación cíclica deberá crear una clase ```class CyclicalEncoding(BaseEstimator, TransformerMixin)``` para la codificación cíclica, por lo que se recomienda ver la [documentación](https://sklearn-template.readthedocs.io/en/latest/user_guide.html).




In [30]:
list(temp_chile_ord.T.index.values)

['Year', 'Month', 'Dates', 'altura', 'latitud', 'longitud', 'station', 'Temp']

In [20]:
#########################
#categoricas
ohe = OneHotEncoder()
cod = ohe.fit_transform(temp_chile_ord.loc[:, ["station"]])
ohe.categories_
#########################

[array(['Futaleufu Ad.', 'Juan Fernandez Estacion Meteorologica',
        'Lautaro Embalse', 'Pichoy Valdivia Ad.', 'Pudahuel Santiago',
        'Santa Juana', 'Santo Domingo Ad.'], dtype=object)]

In [31]:
#min max scaler
minmaxscalercols = ["latitud","longitud","altura","Year"]
minmax_scaler = MinMaxScaler()
minmaxdata = minmax_scaler.fit_transform(temp_chile_ord.loc[:, minmaxscalercols])
minmaxdata = pd.DataFrame(minmaxdata, columns=minmaxscalercols)
minmaxdata

latitud  longitud    altura      Year
0     0.000000  0.790618  0.299357  0.000000
1     0.000000  0.790618  0.299357  0.000000
2     0.627996  0.000000  0.012856  0.035088
3     0.954709  0.925360  0.494949  0.070175
4     0.954709  0.925360  0.494949  0.070175
...        ...       ...       ...       ...
3382  0.626793  0.817600  0.049587  1.000000
3383  0.627996  0.000000  0.012856  1.000000
3384  0.954709  0.925360  0.494949  1.000000
3385  0.644090  0.910425  0.423324  1.000000
3386  1.000000  1.000000  1.000000  1.000000

[3387 rows x 4 columns]

In [32]:
months["Month"]
def ciclico(df):
    return (np.sin((df-1)*(2.*np.pi/12)),np.cos((df-1)*(2.*np.pi/12)))

# se inicializa el transformador
encodingciclico = FunctionTransformer(ciclico)

In [33]:
monthciclo=encodingciclico.transform(months["Month"])

In [ ]:
#class CyclicalEncoding(BaseEstimator,TransformerMixin):
#    def fit(self,X,y=None):
#        return self
#    def transform(self,X):
#        pass   #return np.sin((X-1)*(2.*np.pi/12)) + np.cos((X-1)*(2.*np.pi/12))
    
#months['monthsin'] = np.sin((months["Month"]-1)*(2.*np.pi/12))
#months['monthcos'] = np.cos((months["Month"]-1)*(2.*np.pi/12))
#months

### 2.3.2 Pipeline! [1 Punto]

Realizado el encoding añadan este a un pipeline, anexando como último paso un modelo de regresión. Para el modelo de regresión puede escoger uno de su preferencia o escoger un ```DecisionTreeRegressor()``` o ```LogisticRegression()```. 

Hecho esto, ```Kiguix``` en su rol de querer saber fehacientemente el desempeño de este modelo le sugiere utilizar R2 score, el cual es un estadístico que nos permite visualizar que tan bien predice un modelo de regresión. ¿Que tal le va a la regresión?, ¿Sera coherente utilizar solo estos atributos para predicciones?, Justifique su respuesta.

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(
        temp_chile_ord.drop(columns='Temp'),
        temp_chile_ord.Temp,
        shuffle=True,
        test_size=0.3,
        random_state=42)

##############################

preprocessing_transformer = ColumnTransformer(
    transformers=[
        ('OneHotEncoder', OneHotEncoder(),  ['station']),
        ('MinMaxScaler', MinMaxScaler(),['latitud', 'longitud', 'altura'])
        #,('CodificacionCiclica', function, ['Month'])
    ]
) 

pipeline = Pipeline([('Preprocessing', preprocessing_transformer), ('Modelo', DecisionTreeRegressor())])

##############################

pipeline.fit(X_train, y_train)
predicted_regression = pipeline.predict(X_test)

# Metrica util para medir regresiones
R2score = r2_score(y_test, predicted_regression)

**Justifique:**

```
 Escriba aqui su respuesta
```

### 2.3.3 Predicción del Futuro 🔮 [0,5 Puntos]

En su afán por el proyecto, el don *Dimadon Vallejos, CEO de `Kiguix`* les entrega el siguiente codigo para visualizar las predicciones. ¿Es coherente lo realizado?, comente.

In [ ]:
# Filtramos ciudad
data_santiago = X_train[X_train['station']=='Pudahuel Santiago']
#data_santiago = dataframe_train[dataframe['station']=='Pudahuel Santiago']



# Creamos Información para predecir los proximos años
Prox_anho = data_santiago[data_santiago['Year']==2017][['Year', 'Month','Dates', 'altura', 'latitud',	'longitud', 'station']]
Prox_anho.Year.replace(2017,2018, inplace=True)

# Definimos el regresor entrenado anteriormente
rgr = regresor

# Predecimos
temp_2018 = rgr.predict(Prox_anho)

# Generamos el nuevo DataFrame
temp_2018 = {'Month':data_santiago['Month'].unique(), 'Temp':temp_2018}
temp_2018=pd.DataFrame(temp_2018)
temp_2018['Year'] = 2018
temp_2018

In [ ]:
import plotly.graph_objects as go

forecasted_temp = pd.concat([data_santiago,temp_2018], sort=False).groupby(by='Year')['Temp'].mean().reset_index()
fig = go.Figure(data=[
    go.Scatter(x=forecasted_temp['Year'], y=forecasted_temp['Temp'], mode='lines'),
    go.Scatter(x=forecasted_temp ['Year'], y=forecasted_temp['Temp'], mode='markers')
])
fig.update_layout(title='Temperatura predicha en el tiempo:',
                 xaxis_title='Time', yaxis_title='Grados Celcius')
fig.show()

**Justifique:**

```
 Escriba aqui su respuesta
```

# Cierre
Recuerden que la tarea tiene un plazo de entrega de dos semanas y que **los días de atraso son válidos para las tareas**. Cualquier duda de la tarea, no duden en contactarnos por mail o U-cursos.



<p align="center">
  <img src="https://i.pinimg.com/originals/65/ae/27/65ae270df87c3c4adcea997e48f60852.gif">
</p>